tools/configure/cmdArgs.ml

```diff
@@ -0,0 +1,199 @@
+open Util
+
+(** * Command-line parsing *)
+
+type ide = Opt | Byte | No
+
+type nativecompiler = NativeYes | NativeNo | NativeOndemand
+
+module Prefs = struct
+
+type t = {
+  prefix : string option;
+  interactive : bool;
+  libdir : string option;
+  configdir : string option;
+  datadir : string option;
+  mandir : string option;
+  docdir : string option;
+  arch : string option;
+  natdynlink : bool;
+  coqide : ide option;
+  macintegration : bool;
+  browser : string option;
+  withdoc : bool;
+  byteonly : bool;
+  bin_annot : bool;
+  annot : bool;
+  bytecodecompiler : bool;
+  nativecompiler : nativecompiler;
+  coqwebsite : string;
+  warn_error : bool;
+  dune_profile : string;
+  install_enabled : bool;
+}
+
+end
+
+open Prefs
+
+module Profiles = struct
+
+let default = {
+  prefix = None;
+  interactive = true;
+  libdir = None;
+  configdir = None;
+  datadir = None;
+  mandir = None;
+  docdir = None;
+  arch = None;
+  natdynlink = true;
+  coqide = None;
+  macintegration = true;
+  browser = None;
+  withdoc = false;
+  byteonly = false;
+  bin_annot = false;
+  annot = false;
+  bytecodecompiler = true;
+  nativecompiler =
+    if os_type_win32 || os_type_cygwin then NativeNo else NativeOndemand;
+  coqwebsite = "http://coq.inria.fr/";
+  warn_error = false;
+  dune_profile = "release";
+  install_enabled = true;
+}
+
+let devel state = { state with
+  bin_annot = true;
+  annot = true;
+  warn_error = true;
+  dune_profile = "dev";
+  interactive = true;
+  prefix = Some (Filename.concat (Sys.getcwd ()) "_build_vo/default");
+  install_enabled = false;
+}
+
+let devel_doc = "-annot -bin-annot -warn-error yes"
+
+let get = function
+  | "devel" -> devel
+  | s -> raise (Arg.Bad ("profile name expected instead of "^s))
+
+let doc =
+  "<profile> Sets a bunch of flags. Supported profiles:
+     devel = " ^ devel_doc
+
+end
+
+let get_bool = function
+  | "true" | "yes" | "y" | "all" -> true
+  | "false" | "no" | "n" -> false
+  | s -> raise (Arg.Bad ("boolean argument expected instead of "^s))
+
+let get_ide = function
+  | "opt" -> Opt
+  | "byte" -> Byte
+  | "no" -> No
+  | s -> raise (Arg.Bad ("(opt|byte|no) argument expected instead of "^s))
+
+let get_native = function
+  | "yes" -> NativeYes
+  | "no" -> NativeNo
+  | "ondemand" -> NativeOndemand
+  | s -> raise (Arg.Bad ("(yes|no|ondemand) argument expected instead of "^s))
+
+let prefs = ref Profiles.default
+
+let arg_bool f = Arg.String (fun s -> prefs := f !prefs (get_bool s))
+
+let arg_string f = Arg.String (fun s -> prefs := f !prefs s)
+let arg_string_option f = Arg.String (fun s -> prefs := f !prefs (Some s))
+
+let arg_set f = Arg.Unit (fun () -> prefs := f !prefs)
+
+let arg_ide f = Arg.String (fun s -> prefs := f !prefs (Some (get_ide s)))
+
+let arg_native f = Arg.String (fun s -> prefs := f !prefs (get_native s))
+
+let arg_profile = Arg.String (fun s -> prefs := Profiles.get s !prefs)
+
+(* TODO : earlier any option -foo was also available as --foo *)
```

Comment:
What's the general attitude towards this? Coq seems to do the peculiar thing when it comes to command line flags.

Most other linux programs take `-` for short single-character flags and `--` for long multi-character ones, with the ability to combine the former ones together.

Why doesn't/didn't Coq do it this way? Would it be to disruptive to change?

Reply:
Historical context, I guess it would be quite disruptive indeed; more recent tools use the getops convention.

Reply:
> Why doesn't/didn't Coq do it this way?

Coq started in the 80's. The first versions of `configure` were probably from the early 90's. It could have been changed in the 00's at the time of the version 7.0 rewriting, even more that the developers at this time were strong GNU and Debian supporters, but it did not happen.

Combining one-letter options together in the `ls -la` style would probably be disruptive but supporting both one-dash and double-dash long options (such as both `-profile` and `--profile`) would probably be ok (as long as it is done for all long options).

Reply:
For now, I'm maintaining the status quo in this PR, but indeed, argument handling is a complex topic, more than it looks!